cpp/src/arrow/compute/kernels/vector_array_sort.cc

```diff
@@ -172,6 +173,85 @@ class ArrayCompareSorter {
   }
 };
 
+template <>
+class ArrayCompareSorter<DictionaryType> {
+  struct DictionaryInternal {
+    NullPartitionResult p;
+    const std::shared_ptr<Array>& values;
+    const std::shared_ptr<Array>& indices;
+    const UInt64Array& indices_values;
+    const ArraySortOptions& options;
+    int64_t offset;
+
+    Status Visit(const DataType& index_type) {
+      return Status::TypeError("Dictionary sorting not supported for index type ",
+                               index_type.ToString());
+    }
+
+    template <typename IndexType>
+    enable_if_t<is_integer_type<IndexType>::value, Status> Visit(
+        const IndexType& index_type) {
+      return SortInternal<IndexType>();
+    }
+
+    template <typename IndexType>
+    Status SortInternal() {
+      using ArrayType = typename TypeTraits<IndexType>::ArrayType;
+      using GetView = GetViewType<IndexType>;
+      const auto& indices_array = checked_cast<const ArrayType&>(*indices);
+
+      std::vector<uint64_t> sort_order(indices_values.length());
+      uint64_t cur = 0;
+      auto cur_idx = GetViewType<UInt64Type>::LogicalValue(indices_values.GetView(cur));
+      auto cur_val = values->GetScalar(cur_idx);
+      for (int i = 0; i < indices_values.length(); i++) {
```

Comment:
I don't understand what this loop is doing, can you elaborate?
IIRC, the plan was to first sort the dictionary values, then to get a sort order based on that. I don't see where the dictionary values get sorted...

Reply:
Note that we now have a Rank kernel which should make this easier, for example in Python:
```python
>>> pc.rank(["b", None, "b", "a", None, "c"], tiebreaker="dense")
<pyarrow.lib.UInt64Array object at 0x7ff8a7378dc0>
[
  2,
  4,
  2,
  1,
  4,
  3
]
```

Reply:
This loop is assigning the same label to the same values. And I think I can use rank instead of that loop, I'll go for that change.